### DISCUSSION TITLE
LightHouse - "Remove unused JavaScript" - How to localise unnecessary JS code?

### DISCUSSION THREAD

**Discussion by Vadorequest** [vercel/next.js#13646] [Help]
![image](https://user-images.githubusercontent.com/3807458/83452801-afff2500-a459-11ea-819b-ec2ab8868787.png)

LightHouse report: https://lighthouse-dot-webdotdevsite.appspot.com//lh/html?url=https%3A%2F%2Fnrn-v2-mst-aptd-gcms-lcz-sty-c1-85zxarrqg.now.sh%2Ffr

I tried to use Bundle Analyzer to see what part of the code is unnecessary on the browser, but it's not really helpful, there are many packages listed and they don't really represent the really of the eventual production bundle.

![image](https://user-images.githubusercontent.com/3807458/83452991-f9e80b00-a459-11ea-9324-f7d4d8536944.png)

Above, it the analyser tells me that Font Awesome is > 1Mo, but that's not what's actually bundled in the production build.

I lack observability of what packages really make my production bundle grow.

How do you optimise such things? Does Next plan on providing better analysing tool in a near future? Or improve code splitting? (the LH report shows that only 30%-40% of the `pages/_app.js` is actually used, feels quite a low percentage, why is that?)

**Comment by Vadorequest** [comment]
Just found https://github.com/vercel/next.js/discussions/10655 and will take a deeper look.

**Reply by Vadorequest** [reply]
I tried using `madge`, but it didn't really help me locate unused code.

PR: https://github.com/UnlyEd/next-right-now/pull/82

**Reply by IRediTOTO** [reply]
So what is your final solution? @@ I am searching for this too

**Reply by IRediTOTO** [reply]
with out this site still so fast but I want it 100 point on Google speed. Tell me if you found a good solution :D

**Reply by Vadorequest** [reply]
Haven't found any related to LH. But found `next-unused` to locate source code that isn't imported.

https://github.com/UnlyEd/next-right-now/pull/82

**Reply by pederklev** [reply]
> with out this site still so fast but I want it 100 point on Google speed.

It seems getting to 100 performance for a feature-rich webpage is close to impossible now. I have a test case with a next.js-application using getStaticProps and Preact. So the js overhead is very low in my opinion (First Load JS shared by all: 34.9 kB). The page returns an image which is properly scaled and lazyloaded. With just the image the page has a 100 performance score. Adding an additional `<div><h1>{props.title}</h1><h3>{props.subtitle}</h3></div>` takes the page from 100 to 99. There is a little CSS inside styled-jsx, no frameworks or anything there.

So, even a bare bones case very quickly goes from 100 to 99. I'd rather aim for >90, I think.

**Reply by IRediTOTO** [reply]
> > with out this site still so fast but I want it 100 point on Google speed.
> 
> It seems getting to 100 performance for a feature-rich webpage is close to impossible now. I have a test case with a next.js-application using getStaticProps and Preact. So the js overhead is very low in my opinion (First Load JS shared by all: 34.9 kB). The page returns an image which is properly scaled and lazyloaded. With just the image the page has a 100 performance score. Adding an additional `<div><h1>{props.title}</h1><h3>{props.subtitle}</h3></div>` takes the page from 100 to 99. There is a little CSS inside styled-jsx, no frameworks or anything there.
> 
> So, even a bare bones case very quickly goes from 100 to 99. I'd rather aim for >90, I think.

Yes, >=90 is ok. Problem just come from unused JS  and my stupid when code :D 
![image](https://user-images.githubusercontent.com/15695252/94140240-24d8ed80-fe95-11ea-8e0d-85ff2b2954c5.png)

**Reply by IRediTOTO** [reply]
Guys, do you use CSS in JS? Is it make site slow by inline css?

**Reply by ljosberinn** [reply]
@IRediTOTO none of the commonly used css-in-js solutions should have a noticeble performance impact.

@pederklev that is odd, as I can reach perf of 100 on a css-in-js heavy site. Please keep in mind that Lighthouse should run in a Guest environment in Chrome, as browser extensions, in this case most notably the React DevTools, will slow down the test process. Besides that, Lighthouse is dependant on your machines CPU. It's not an external audit. So you will see varying results depending on machine and opened apps etc.

**Reply by pederklev** [reply]
@ljosberinn I run Lighthouse in an incognito window to avoid most extensions. For deployed sites i crosscheck my local results against https://developers.google.com/speed/pagespeed/insights/ , and rarely find any major differences. Do you know of any webpages* which can reach a 100 performance score on mobile, that you can show me? I haven't seen any since the change in how Lighthouse weights its scores a few months ago.

*Beyond simple blogs and the likes.

**Reply by IRediTOTO** [reply]
I got 78  on mobile just 2 more point is ok. :) 
![image](https://user-images.githubusercontent.com/15695252/94287785-ec164280-ff80-11ea-852f-36eb314cfc59.png)

**Comment by prateekbh** [comment]
@Vadorequest is the repo open source by any chance? why do u think the bundle analyzer output is incorrect? would love to verify that if can look at the code

**Reply by Vadorequest** [reply]
Yes it is!

https://github.com/UnlyEd/next-right-now/tree/v2-mst-aptd-at-lcz-sty

Quick start: https://unlyed.github.io/next-right-now/getting-started/quick-start

Well, the main issue is observability. Even though LH tells me here is a lot of unused code, I don't know how to locate it. FortAwesome shouldn't take 1Mo, for example.

**Reply by prateekbh** [reply]
do i still run `ANALYZE=true yarn build`? cuz i cant get analyzer to start

**Reply by prateekbh** [reply]
nvm found it

**Comment by prateekbh** [comment]
@Vadorequest the reason why your bundle analyzer is off is because you're running `ANALYZE_BUNDLE=true yarn start`.
This runs the analyzer against the dev server bundles which do not have production optimizations like tree shaking/ minification etc. You need to replace it with `ANALYZE_BUNDLE=true next build`

**Reply by Vadorequest** [reply]
Thanks for this precision!

I added those scripts to better reflect what should be done:
    "analyse:bundle:development": "ANALYZE_BUNDLE=true yarn start",
    "analyse:bundle:production": "ANALYZE_BUNDLE=true next build",

I believe it'll improve DX.

But, it doesn't answer the initial issue about observability. (what part of JS is bundled but not being used) Even though it'll help observe what's really bundled in production.

**Comment by LuizHAP** [comment]
Fellas, first of all, I thank you for the conversation that you are having about this, and I am really more and more following what you are doing. 
Just to complement, I also have a big ecommerce project, and I can't get past the score 10 on mobile (almost 95% of my access is mobile). As it is a private code, I cannot share it with you, but I have tried in every way to use dynamic too, but during my bundle, it continues to import and create .js files for components that are not even being used. 

Are you thinking of any step by step to improve the score of sites that use CSS frameworks, such as Material-UI, to improve the score?

![image](https://user-images.githubusercontent.com/20971507/94300707-c6b51480-ff3f-11ea-89f0-e400b51e8783.png)
![image](https://user-images.githubusercontent.com/20971507/94300731-d0d71300-ff3f-11ea-87ed-d355e54dc7b0.png)

Just to complete, I am new to using nextjs, and I realize that most of the bundle that is generated depends a lot on the components that you are using, but in a large project, which is almost finished, it is complicated for you to change the entire structure and components , which in my case is to remove the Material-UI, for example.

I realize that in the bundle, the biggest consumer is a third-party, .js customer service file on my website. What is your tip for this type of case?

**Reply by prateekbh** [reply]
@LuizHAP can you share the URL?

**Reply by rokinsky** [reply]
@prateekbh https://prototype-staging.uaubox.com.br

**Reply by IRediTOTO** [reply]
I got this before, but it from your code, not from Nextjs. Let's try to find it

**Reply by GuiPerim** [reply]
So Luiz, i facing the same problem that you. How did you improve your unused JS? I also pretty new on NextJs

**Comment by tania-asimi** [comment]
Hello everyone,
I'm getting pretty much the same issue. When I'm trying to find out why I keep get that warning to 'Remove unused Javascript' and in DevTools I'm looking at the big chunks using 'Show Coverage', I can see that for one page, all the other components from other pages are being loaded in next.js chunks as well, and I am using the dynamic imports, but apparently it doesn't matter that much.  
Anyone found a solution so far?

**Reply by skworden** [reply]
If you are using lighthouse in the console, disable dev tools and re-run lighthouse. This will get rid of a few of the unused script warnings to help isolate the problem.

Also, any of the unused scripts xxx.../some_page_on_your_site.1234.js are being loaded by next/link.  It prefetches links that are in the viewport.  You can disable this by setting prefetch to false on links that users will not likely click on first.  This will remove the scripts for prefetched pages.

You will also want to run it against a next build and not in dev mode.  This will also remove some unused dev scripts.

As far as some of the chunks... I have no clue.  I only have one unused script on my site and I don't have a clue what it's from.  I have checked for unused code manually and with next-unused...I can't find any. I can't figure out what it's for.

I'd also personally like next to bundle all of the scripts used on every page (identified during build) into one file to reduce the number of request.

**Reply by JustFly1984** [reply]
I can say that I can track some unused code in forks for mobile/desktop views, then some component loads in case screen width is less than some value, and another loads of screen width is more than some value in some me projects. Also could be some extra code, if you use svg as components with base64 images embedded.

**Reply by skworden** [reply]
Figured part of my other chunk... It is a few things bunded together.

I had `import { throttle } from 'lodash'` which was bundling all of lodash.  I changed it to `import throttle from 'lodash/throttle' ` and this removed the unused lodash from the chunk. 

I also use framer motion and it also appears to be loading the whole package but I can't figure out what to do about it.

**Reply by JustFly1984** [reply]
Write an issue to framer motion github repository.
Better use redux-debounce, or ts-debounce instead of lodash throttle, as lodash has security issues all the time.

**Reply by skworden** [reply]
In the latest version of framer motion they have addressed the issue in my last comment https://www.framer.com/api/motion/guide-reduce-bundle-size/ 

It appears that most of the unused JS (lighthouse is calling out) that is getting built by next is generated from third party libraries importing their whole or unused parts of their library and/or how you export/import your own code.
See previous comment for a simple example.

I sometimes get a framework.js file showing up in the unused js list but I haven't figured out a fix for that.

I found my unused code by opening the unused code chunk file and looked for unique function names (the code is nearly unreadable because it's minified but library function names will remain untouched).

I googled the function names and found the libraries causing the issue - then worked backwards to figure out how to fix the importing issues.  It might be difficult to fix all of the issues for third-party libraries because they would have to fix them.

I've spent like 5 hours on this and I'm taking a pretty consistent 99, sometimes 100, and randomly 96+ (maybe lambda cold start/cache miss? idk?) performance on mobile for my static next site as a W and stopping this madness of trying to get a 100 every time.

**Reply by Vadorequest** [reply]
Good tip @lifedup. I'd add that instead of googling it, looking for them within `node_modules` is very likely to be more efficient.

**Reply by skworden** [reply]
That makes more sense and would probably be for efficient, I was being lazy - I hide node_modules from vscode and don't know of a good way to search them without changing my vscode setup.

**Reply by Vadorequest** [reply]
I don't know how to do that, I use WebStorm and even though node_modules is "ignored", it's still searchable easily by clicking on it then cmd+shift+f.

**Reply by skworden** [reply]
Here is a article to help you get started tree shaking to help identify culprits - https://medium.com/ne-digital/how-to-reduce-next-js-bundle-size-68f7ac70c375

**Comment by JustFly1984** [comment]
As I understand React based frameworks, in context of unused JavaScript analysed by the lighthouse, a lot of it is ‘IF’ or ‘Switch’ case branches based on media screen sizes. Then you have some content hidden for mobile, or desktop, it does execute one of the branches, leaving other branch as unused JavaScript. This is why you get different results in kB for mobile and desktop tests.

**Reply by skworden** [reply]
That code determining what to render is being used and would not be considered in this.   This is caused by how libraries are exporting /importing their code.  Some libraries will export all of their code in one export regardless of what you are using.  The only way to identify it is to do tree shaking and process of elimination.  Your likely not going to be able to get to zero but you can identify libraries that are killing your performance.   Here is a common library used for this problem - https://www.npmjs.com/package/webpack-bundle-analyzer

**Comment by ponikar** [comment]
![image](https://user-images.githubusercontent.com/54766861/130430419-979e06cb-cf1d-4f9f-ab9a-eaac94832ddd.png)

I am having the same issue. When I opened the network tab I discovers lots of JS files which I am not aware of it some of them. Although After reading above comments, some of them started making sense.

Still I wonder any good solution to removed this unused files.

**Reply by skworden** [reply]
Someone else had this question on [stack](https://stackoverflow.com/a/68642023/15304814) I tried to answer - you might be able to get some more insight there.

**Reply by ponikar** [reply]
Can you tell me what is the role of _ssgManifest.js and buildManifest.js?

**Reply by skworden** [reply]
I don't know how Next uses them but the appear to be internal instructions / file mapping for built projects.

**Reply by ponikar** [reply]
Okay.

**Comment by Ankur-2002** [comment]
![image](https://user-images.githubusercontent.com/76253181/137990647-5d77eecb-356e-4e4b-b5b4-7d7993413138.png

Can anyone help me? I want to decrease the initial loading in my project but these are giving very huge. So, is there any solution to delete this and accomplish that task?

**Comment by mlnima** [comment]
to find unused js I run lighthouse on development mode in the Incognito tab, it will tell the exact component causing this error. but, when I optimize almost everything and use dynamic imports even when I have 2 different components for Mobile and desktop(you can detect from the user agent inside getServerSideProps), it will show unused Js on the mobile test. I did my best nothing is imported or calculated extra.
The desktop version is perfect but on mobile, I see this error on Next-image absolutely for no reason from my component.
Today after updating to React18 and NextJs 2.1.4 my first load increased from 152 to 172kb

**Comment by gaspergigio** [comment]
You solve it by adding the following in the nextconfig:

```
pages: {
        '*': {
          maxChunkSize: 30000
        },
    },
```

With this each chunk cannot exceed 30KB. Being smaller and having more chunks, they will load optimally as a library is demanded by the app.

**Reply by catchellers** [reply]
what is property pages in next.config? I just read the docs nextjs about this and not see anything about "pages"

**Reply by realestninja** [reply]
I would also like to know where this `maxChunkSize` option is coming from, as I can't find any documentation on this. 
Is this a suggestion that was hallucinated by some AI?

**Reply by javad7z7** [reply]
use this 
` webpack: (config, { isServer }) => {
    if (!isServer) {
      config.optimization.splitChunks.maxSize = 30000;
    }
    return config;
  },`
